apps/settings/main.py

```diff
@@ -1,236 +1,158 @@
+from __future__ import division
 
 import os
-import signal
-from subprocess import check_output
+import shlex
+import shutil
+import subprocess
+import tempfile
 from time import sleep
 
-try:
-    import httplib
-except:
-    import http.client as httplib
-
-# Using a TextProgressBar because only it shows a message on the screen for now
-from ui import Menu, PrettyPrinter, DialogBox, TextProgressBar, Listbox
+from apps import ZeroApp
+from helpers.general import ZPUI_INSTALL_DIR
 from helpers.logger import setup_logger
+from ui import Menu, DialogBox, Printer, ProgressBar
 
-menu_name = "Settings"
 logger = setup_logger(__name__, "info")
 
 
-class GitInterface():
-
-    @classmethod
-    def git_available(cls):
-        try:
-            cls.command("--help")
-        except OSError:
-            return False
-        return True
-
-    @staticmethod
-    def command(command):
-        commandline = "git {}".format(command)
-        logger.debug("Executing: {}".format(commandline))
-        return check_output(commandline, shell=True)
-
-    @classmethod
-    def get_head_for_branch(cls, branch):
-        output = cls.command("rev-parse {}".format(branch)).strip()
-        return output
-
-    @classmethod
-    def get_current_branch(cls):
-        return cls.get_head_for_branch("--abbrev-ref HEAD").strip()
-
-    @classmethod
-    def checkout(cls, reference):
-        return cls.command("checkout {}".format(reference))
-
-    @classmethod
-    def pull(cls, source = "origin", branch = "master", opts="--no-edit"):
-        return cls.command("pull {2} {0} {1}".format(source, branch, opts))
-
-
-class UpdateUnnecessary(Exception):
-    pass
-
-
-class GenericUpdater(object):
-    steps = []
-    progressbar_messages = {}
-    failed_messages = {}
-
-    def run_step(self, step_name):
-        logger.info("Running update step: '{}'".format(step_name))
-        getattr(self, "do_" + step_name)()
-        logger.debug("Update step '{}' completed!".format(step_name))
-
-    def revert_step(self, step_name):
-        if hasattr(self, "revert_" + step_name):
-            logger.info("Reverting update step: '{}'".format(step_name))
-            getattr(self, "revert_" + step_name)()
-            logger.debug("Update step '{}' reverted!".format(step_name))
+class ZpuiUpdaterApp(ZeroApp):
+    def __init__(self, i, o):
+        super(ZpuiUpdaterApp, self).__init__(i, o)
+        self.menu_name = "Update ZPUI"
+        self.menu = Menu(
+            [
+                ["Update ZPUI", self.update_zpui],
```

Comment:
I don't think having separate menu options is a good idea.
1) as we add more possible branches for users to try out and, possibly, use, we'd need to add (and hardcode! ) more menu entries
2) the user is supposed to remember which branch they're on before updating, so the update is no longer as straightforward as "press this button"
3) a misclick will make the user update from a wrong branch, which will lead in 1) long update times 2) possible problems during update if the branch they're updating from is not tested enough. Basically, switching ZPUI branches has to be a conscious choice.